src/v/archival/async_data_uploader.cc

```diff
@@ -0,0 +1,537 @@
+            bool max_bytes_reached = _parent->_buffer.size_bytes()
+                                     > (size_t)_parent->_max_bytes;
+            vlog(
+              _parent->_ctxlog.debug,
+              "max_bytes_reached: {}",
+              max_bytes_reached);
+            co_return max_bytes_reached ? ss::stop_iteration::yes
+                                        : ss::stop_iteration::no;
+        }
+
+        bool end_of_stream() const { return false; }
+
+    private:
+        reader_ds* _parent;
+        inclusive_offset_range _range;
+    };
+
+    friend class consumer;
+
+public:
+    explicit reader_ds(
+      model::record_batch_reader r,
+      size_t max_bytes,
+      inclusive_offset_range range,
+      model::timeout_clock::time_point deadline,
+      ss::gate::holder h,
+      retry_chain_logger& rtc)
+      : _reader(std::move(r))
+      , _max_bytes((ssize_t)max_bytes)
+      , _deadline(deadline)
+      , _range(range)
+      , _holder(std::move(h))
+      , _ctxlog(rtc) {}
+
+    ss::future<ss::temporary_buffer<char>> get() override {
+        // Consume using the reader until the reader is done
+        if (_buffer.empty() && !_reader.is_end_of_stream()) {
+            vlog(
+              _ctxlog.trace, "Buffer is empty, pulling data from the reader");
+            consumer c(this, _range);
+            auto done = co_await _reader.consume(c, _deadline);
+            vlog(
+              _ctxlog.trace,
+              "Consume returned {}, buffer has {} bytes",
+              done,
+              _buffer.size_bytes());
+        }
+        if (_buffer.empty()) {
+            vlog(
+              _ctxlog.trace,
+              "Reader end-of-stream: {}",
+              _reader.is_end_of_stream());
+            ss::temporary_buffer<char> empty;
+            co_return empty;
+        }
+        auto top = _buffer.begin();
+        auto buf = std::move(*top).release();
+        _buffer.pop_front();
+        vlog(
+          _ctxlog.debug,
+          "get produced {} bytes, {} bytes buffered",
+          buf.size(),
+          _buffer.size_bytes());
+        co_return buf;
+    }
+    ss::future<ss::temporary_buffer<char>> skip(uint64_t n) override {
+        ss::temporary_buffer<char> empty;
+        co_return empty;
+    }
+    ss::future<> close() override {
+        _holder.release();
+        co_await std::move(_reader).release()->finally();
+    }
+
+private:
+    model::record_batch_reader _reader;
+    iobuf _buffer;
+    ssize_t _max_bytes;
+    size_t _num_batches{};
+    model::timeout_clock::time_point _deadline;
+    inclusive_offset_range _range;
+    ss::gate::holder _holder;
+    retry_chain_logger& _ctxlog;
+};
+
+ss::input_stream<char> make_reader_input_stream(
+  model::record_batch_reader r,
+  size_t read_buffer_size,
+  inclusive_offset_range range,
+  model::timeout_clock::duration timeout,
+  ss::gate::holder h,
+  retry_chain_logger& rtc) {
+    auto deadline = model::timeout_clock::now() + timeout;
```

Comment:
I'm curious why we are mixing timeouts (which are propagated down to very fine grained operations) with deadlines?